### PR TITLE
CI maintenance updates and adding CI Unit Tests

### PR DESCRIPTION
**Pull request by TerrenceMcGuinness-NOAA**
# Description

This PR has a few maintenance updates to the CI pipeline and adds a test directory with Unit Tests

**Major Maintenance updates:**

- Added try blocks with appropriate messaging to GitHub PR of failure for:
- - **scm** checkout
- - build fail (with error logs sent as gists)
- - create experiment fails with `stderr` sent to GitHub PR messaging
- Pre-stage FAILS from the above are now captured these fails allow FINALIZE to update the label to FAIL (i.e. no more "hanging" CI state labels in GitHub - see image below)

**Minor Maintenance updates:**
- Fix for STALLED cases reviled from PR 2700 (just needed a lambda specifier)
- Fixed path to experiment directory in PR message (had dropped EXPDIR in path)
- Needed `latin-1` decoder in reading log files for publishing

**Added python Unit Tests for CI functionality:**
- Installed **Rocoto** and **wxfow** in GitHub Runner for testing key CI utility codes
- Cashed the install of Rocoto in the GitHub Runners to greatly reduce stetup time for running the unit tests
- Unit Tests Python scripts added
  - `test_rocostat.py`: rocoto_statcount() rocoto_summary() rocoto_stalled()
  - `test_setup.py`: setup_expt() test_setup_xml()
  - `test_create_experment`: test_create_experiment()
  - - Runs all PR cases that do not have ICs in the GItHub Runner
  -  Reporting mechanism in the Actions tab for Python Unit Testing results
  - Test case data for STALLED and RUNNING stored on S3 and pulled using wget during runtime of tests

- Bug fix (fixes something broken)
- Maintenance (new CI unit tests)

# How has this been tested?

Unit Tests [ran](https://github.com/TerrenceMcGuinness-NOAA/global-workflow/actions/runs/9716954083) in GitHub Runner using `.github/workflow/ci_uint_tests.yaml`
```
test_create_experiment.py::test_create_experiment PASSED                 [ 12%]
test_rocotostat.py::test_rocoto_statcount PASSED                         [ 25%]
test_rocotostat.py::test_rocoto_summary PASSED                           [ 37%]
test_rocotostat.py::test_rocoto_done PASSED                              [ 50%]
test_rocotostat.py::test_rocoto_stalled PASSED                           [ 62%]
test_setup.py::test_setup_expt PASSED                                    [ 75%]
test_setup.py::test_setup_xml PASSED                                     [ 87%]
test_setup.py::test_setup_xml_fail_config_env_cornercase PASSED          [100%]
```

image (1)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

### CODE REVIEW

File: .github/workflows/ci_uint_tests.yaml
Comment (DavidHuber-NOAA) [inline .github/workflows/ci_uint_tests.yaml:1]: Should be ci_unit_tests.yaml.

File: .github/workflows/ci_uint_tests.yaml
Comment (DavidHuber-NOAA) [inline .github/workflows/ci_uint_tests.yaml]: Should we be using the same Python version as spack-stack (3.11.6)?

File: ci/cases/yamls/gefs_ci_defaults.yaml
Comment (DavidHuber-NOAA) [inline ci/cases/yamls/gefs_ci_defaults.yaml:4]: Should `ACCOUNT_SERVICE` also get set here?  If so, this should probably be expanded to the other CI yamls.

File: ci/scripts/tests/test_create_experiment.py
Comment (DavidHuber-NOAA) [inline ci/scripts/tests/test_create_experiment.py]:
```suggestion
    create_experiment_script = Executable('../../../workflow/create_experiment.py')
    create_experiment_script.add_default_arg(['-y', f'../../cases/pr/{case}.yaml', '--overwrite'])
    create_experiment_script(env=env)
    assert (create_experiment_script.returncode == 0)
```

File: ci/scripts/tests/test_setup.py
Comment (DavidHuber-NOAA) [inline ci/scripts/tests/test_setup.py:10]: Since `fv3-cpu` is the default on Hera, perhaps another account should be used to verify it is being set correctly.  I, for one, do not have access to `fv3-cpu` and instead use `nems`.  Also, perhaps `account_service` should be set here to something different than `account`.

File: ci/scripts/tests/test_setup.py
Comment (DavidHuber-NOAA) [inline ci/scripts/tests/test_setup.py:37]: Also check `ACCOUNT_SERVICE`:
```suggestion
    assert base.ACCOUNT == account
    assert base.ACCOUNT_SERVICE == account_service
```

File: ci/scripts/tests/test_setup.py
Comment (DavidHuber-NOAA) [inline ci/scripts/tests/test_setup.py]: To improve tabbing:
```suggestion
    script_content = (`#!/usr/bin/env bash\n`
                      `export HOMEgfs=foobar\n`
                      `../../../workflow/setup_xml.py "${1}"\n`)
```

File: ci/scripts/tests/test_setup.py
Comment (DavidHuber-NOAA) [inline ci/scripts/tests/test_setup.py]:
```suggestion
        assert "UNKNOWN" not in base.values()
```

File: ci/scripts/tests/test_setup.py
Comment (DavidHuber-NOAA) [inline ci/scripts/tests/test_setup.py:81]: Is the intent here to pass the test or fail it?  The comment says it should pass, but `pytest.fail` is being called.  If it is intended to fail, then I suggest:
```suggestion
        # If an unexpected exception occurs, fail the test with a custom message
        pytest.fail(f"Unexpected exception occurred: {e}")
```

**Review by DavidHuber-NOAA — CHANGES_REQUESTED** [review]

File: ci/scripts/tests/test_setup.py
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/tests/test_setup.py:10]: No this is a static test from staged input data.

File: ci/scripts/tests/test_setup.py
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/tests/test_setup.py:37]: Account was chosen here because it once was testing **wxflow** name clashing.  A single test of **config.base** is sufficient, adding ACCOUNT_SERVICE is arbitrary.

File: ci/scripts/tests/test_setup.py
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/tests/test_setup.py]: I thought I fixed that.

File: ci/scripts/tests/test_setup.py
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/tests/test_setup.py:81]: No this is an "expected" exception, ridicules as it sounds this test is making sure it fails.  This is a known feature bug.  It is put here to point that out.

File: ci/cases/yamls/gefs_ci_defaults.yaml
Comment (TerrenceMcGuinness-NOAA) [reply ci/cases/yamls/gefs_ci_defaults.yaml:4]: @aerorahul Not sure if we need to over ride this one too.

Comment (aerorahul) [reply ci/cases/yamls/gefs_ci_defaults.yaml:4]: It seems like `ACCOUNT_SERVICE => ACCOUNT`.  It was added in #2627.  Looking at all the yamls in `workflow/hosts/`, `ACCOUNT_SERVICE == ACCOUNT` and should not be necessary.
@guoqing-noaa Do you remember the nuance?
Is there ever an instance where `ACCOUNT_SERVICE` is not the same as `ACCOUNT`?

File: ci/scripts/tests/test_setup.py
Comment (aerorahul) [inline ci/scripts/tests/test_setup.py]:
```suggestion
    assert "UNKNOWN" not in base.values()
```

File: .github/workflows/ci_uint_tests.yaml
Comment (TerrenceMcGuinness-NOAA) [reply .github/workflows/ci_uint_tests.yaml:1]: lol

File: .github/workflows/ci_uint_tests.yaml
Comment (TerrenceMcGuinness-NOAA) [reply .github/workflows/ci_uint_tests.yaml]: I'm going to so no.  .002 ahead in python should not impact these tests.  This version is not arbitrary and it was difficult to land on on a usable number for the image running on the Runner in the first place.

File: ci/scripts/utils/publish_logs.py
Comment (DavidHuber-NOAA) [inline ci/scripts/utils/publish_logs.py:50]: `file` is being redefined within the loop.

File: ci/scripts/utils/publish_logs.py
Comment (DavidHuber-NOAA) [inline ci/scripts/utils/publish_logs.py:90]: Same here.

File: ci/scripts/utils/publish_logs.py
Comment (TerrenceMcGuinness-NOAA) [reply ci/scripts/utils/publish_logs.py:50]: I don't understand the issue with this?  Each file is opened to a new file object with this specific encoding an is read read.

Comment (DavidHuber-NOAA) [reply ci/scripts/utils/publish_logs.py:50]: You're right, it doesn't matter since you aren't trying to work with the original file object.  Resolving.

**Comment by TerrenceMcGuinness-NOAA** [comment]
`publish_logs.py` failed to authenticate to GitHub on Hercules.  So this FAIL is not a CI PASS. It should have given us the message with the error log as gist.  Looking into this now.

**Review by DavidHuber-NOAA — DISMISSED** [review]
Approved pending successful completion of CI.

**Comment by TerrenceMcGuinness-NOAA** [comment]
After the pipeline correctly identified and located and reported the correct error log from a build fail, it exposed an error in the logic of the pipeline to run Finalize and update the label to FAIL.  Working on this discrepancy in this PR now.

Updated the label to CI-Hera-Failed here by hand so the actual build fail can also be addressed.

**Comment by DavidHuber-NOAA** [comment]
The `C48_S2SWA_gefs` test failed at the `create_experiment` step on Hercules as `CLUSTERS` is not defined in the task_config.  I'll add another hot fix.

**Comment by DavidHuber-NOAA** [comment]
The `create_experiment` issue was resolved by #2748.

File: ci/scripts/tests/test_create_experiment.py
Comment (aerorahul) [inline ci/scripts/tests/test_create_experiment.py]:
```suggestion
    create_experiment_script = Executable(f'{HOMEgfs}/workflow/create_experiment.py')
    create_experiment_script.add_default_arg(['-y', f'{HOMEgfs}/ci/cases/pr/{case}.yaml', '--overwrite'])
```
Use `HOMEgfs` declared on line 6 instead of relative paths

**Review by aerorahul — DISMISSED** [review]
Just one comment.

**Comment by DavidHuber-NOAA** [comment]
It looks like Hera didn't pick up the label.

**Comment by TerrenceMcGuinness-NOAA** [comment]
It did get picked up but in that last integration before I left there was a syntax error in the pipeline so it doesn't run.

**Comment by TerrenceMcGuinness-NOAA** [comment]
Needs "cluster" update and apply catch fix similar to build fail to Finalize in the Pipeline.

```
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/CI/2740/gefs//workflow/create_experiment.py", line 101, in <module>
    setup_xml.main(setup_xml_args)
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/setup_xml.py", line 73, in main
    xml = rocoto_xml_factory.create(f'{net}_{mode}', app_config, rocoto_param_dict)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/wxflow/factory.py", line 71, in create
    return self._builders[key](*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/rocoto/gefs_xml.py", line 14, in __init__
    super().__init__(app_config, rocoto_config)
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/rocoto/workflow_xml.py", line 27, in __init__
    task_list = get_wf_tasks(app_config)
                ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/rocoto/workflow_tasks.py", line 21, in get_wf_tasks
    tasks.append(task_obj.get_task(task_name))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/rocoto/tasks.py", line 257, in get_task
    return getattr(self, task_name, *args, **kwargs)()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/rocoto/gefs_tasks.py", line 72, in stage_ic
    resources = self.get_resource('stage_ic')
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/rocoto/tasks.py", line 2[29](https://jenkins.epic.oarcloud.noaa.gov/job/global-workflow/job/EMC-Global-Pipeline/view/change-requests/job/PR-2740/16/pipeline-console/?start-byte=0&selected-node=250#log-29), in get_resource
    if task_config['CLUSTERS'] not in ["", '@CLUSTERS@']:
       ~~~~~~~~~~~^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/CI/2740/gefs/workflow/wxflow/attrdict.py", line 84, in __missing__
    raise KeyError(name)
KeyError: 'CLUSTERS'
script returned exit code 1
```

**Comment by DavidHuber-NOAA** [comment]
@TerrenceMcGuinness-NOAA Is this ready to be re-reviewed and tested?

**Review by aerorahul — APPROVED** [review]
looks good.
A self-test should be run on develop ASAP after this is merged.

**Pull request merged**
